### ISSUE TITLE
`resolveBrowserLocale` not returning full locale

### ISSUE THREAD

**Issue by FernandoKGA** [marmelab/react-admin#7170]
**What you were expecting:**

When using `resolveBrowserLocale`, it would get the correctly locale and translate to it.

**What happened instead:**

The function returned only the first part of the locale. In my case, the locale is **"pt-BR"**, which returned as **"pt"** only.

**Steps to reproduce:**

1. Use `resolveBrowserLocale` at `polyglotI18nProvider` `initialLocale` paremeter or other function that receives a locale from `resolveBrowserLocale`.
2. Use a language at the browser that has a region locale (example: `pt-BR`).
3. Create a translation file for the locale.
4. Observe that the translation won't work if you use the string with region.

**Related code:**

https://codesandbox.io/s/ecstatic-clarke-ksfqx?file=/src/i18nProvider.tsx:354-374

**Other information:**

Reporting this problem because when your application deals with regions from the same language ("pt-PT" and "pt-BR"), this function cannot be used.

**Suggestions:**
- Removing the split from [TranslationUtils.ts](https://github.com/marmelab/react-admin/blob/b71d7ddb3d7ad3379ed97819ac1cb4abe38ca74e/packages/ra-core/src/i18n/TranslationUtils.ts#L39), but this can cause many bugs across applications.
- Inserting an option to allow full string from locale to be return at [TranslationUtils.ts](https://github.com/marmelab/react-admin/blob/b71d7ddb3d7ad3379ed97819ac1cb4abe38ca74e/packages/ra-core/src/i18n/TranslationUtils.ts#L39)

```js
export const resolveBrowserLocale = (defaultLocale = DEFAULT_LOCALE, allowFullLocale = false) => {
    // Rely on the window.navigator object to determine user locale
    const {
        language,
        browserLanguage,
        userLanguage,
    } = window.navigator as AllNavigatorLanguage;
    const language = (language || browserLanguage || userLanguage || defaultLocale);
    return allowFullLocale ? language : language.split('-')[0];
```

**Environment**

* React-admin version: 4.0.0-alpha.2 (paired with master)
* React version: `^17.0.0`
* Browser: Firefox 96.0 (64-bits) Mozilla Firefox for Ubuntu canonical - 1.0

**Comment by FernandoKGA** [comment]
So... I've been messing around with the function and we can improve it in two ways:

1. Adding an another default parameter, not breaking the function interface, and not becoming a breaking change.
2. Shifting to [destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment), but this will become a breaking change as it will need a object to be supplied to the function.

In the first case, to call the function `resolveBrowserLocale` expressed by this *new* declaration: 
```js
export const resolveBrowserLocale = (
  defaultLocale: string = DEFAULT_LOCALE,
  allowFullLocale: boolean = false
) => {...}
```
you will need to provide `undefined` to the first parameter, following with `true` if you want to allow the return of the full locale string: `resolveBrowserLocale(undefined, true)`.

At the last case, to call the function `resolveBrowserLocale` expressed by this *new* declaration:
```js
export const resolveBrowserLocale = (
    {
        defaultLocale = DEFAULT_LOCALE,
        allowFullLocale = false,
    }: { defaultLocale?: string; allowFullLocale?: boolean } = {
        defaultLocale: DEFAULT_LOCALE,
        allowFullLocale: false,
    }
) => {
```
you will need to provide an object with the property (or properties) that you want to supply the function.

Examples:
```js
resolveBrowserLocale({ defaultLocale: 'en-US' })
resolveBrowserLocale({ allowFullLocale: true })
resolveBrowserLocale({ defaultLocale: 'en-US', allowFullLocale: true })
```

However, as I said, the last one will become a breaking change.

**Comment by fzaninotto** [comment]
Fixed by #7193
